Experimental Schemas: Add support for default values fill in

joaopapereira opened the pull request:
When a value is missing in the data file but present in the schema, ytt
will use this value as the default and it will populate the map
structure with it.

pivotaljohn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -63,6 +63,81 @@ rendered: true`
+`
+	templateYAML := `#@ load("@ytt:data", "data")
+---
+rendered: true
+db:
+  hostname: #@ data.values.db_conn.hostname
+  port: #@ data.values.db_conn.port
+  username: #@ data.values.db_conn.username
+  password: #@ data.values.db_conn.password
+  tls_only: #@ data.values.db_conn.tls_only
+  metadata: #@ data.values.db_conn.metadata
+  other_key: #@ data.values.db_conn.other_key.missing_key.something
+  inner_key: #@ data.values.db_conn.other_key.inner_key
```
Consider taking advantage of the fact that `structs` can be rendered, as is:

```yaml
#@ load("@ytt:data", "data")
---
data.values: #@ data.values
```

joaopapereira replied:
done

pivotaljohn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -63,6 +63,81 @@ rendered: true`
 }
 
+func TestMapOnlySchemaFillInDefaults(t *testing.T) {
+	schemaYAML := `#@schema/match data_values=True
+---
+db_conn:
+  hostname: server.example.com
+  port: 0
+  username: someuser
+  password: somepassword
+  tls_only: false
+  metadata:
+    run: jobName
+  other_key:
+    inner_key: some string
+    missing_key:
+      something: other
+`
+	dataValuesYAML := `#@data/values
+---
+db_conn:
+  username: sa
+  password: changeme
+  tls_only: true
+  other_key:
+    missing_key:
+      something: yeet
+`
+	templateYAML := `#@ load("@ytt:data", "data")
```
Consider shaping the test fixture so that it strikes a balance between using real-worldish data values (which is present, here), and having elements that are telling of what's happening.

Otherwise, the reader must digest the whole example, line-by-line and build up a mental mapping of what's being tested ==> "too" much cognitive load.

For example:
- `missing_key` is not missing, but `inner_key` _is_ missing (from the data values).
- there appear to be multiple places where defaults are overridden in the data values. consider to limit this to — at most — one per type? (both `username` and `password` are defaulted strings that are overridden.

joaopapereira replied:
done

pivotaljohn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -63,6 +63,81 @@ rendered: true`
+  other_key: yeet
+  inner_key: some string
+`
+	if string(out.Files[0].Bytes()) != expected {
+		t.Fatalf("Expected output to only include template YAML, but got: %s", out.Files[0].Bytes())
+	}
```
An emerging pattern is to use our text diffing library to make it easier to spot the differences between expected and actual.

Let's pull that in here.

joaopapereira replied:
done

pivotaljohn commented on pkg/yamlmeta/type.go:
```diff
@@ -103,18 +105,48 @@ func (t *MapType) AssignTypeTo(typeable Typeable) (chk TypeCheck) {
+	}
+}
+
+func contains(haystack []interface{}, needle interface{}) bool {
+	for _, key := range haystack {
+		if key == needle {
+			return true
+		}
+	}
+	return false
+}
```
This made me smile. Thanks for the tongue-in-cheek idiom! ❤️ 

Does `Key`s are technically any type. Does `==` do a deep equals? 🤔

joaopapereira replied:
Key should be a string (or perhaps a number???) which should render this `==` to be a simple string comparison.
This method is already used in other parts of the code to ensure we are handly the value from the correct key. (type.go:112 as an example)

pivotaljohn commented on pkg/yamlmeta/type.go:
```diff
@@ -48,8 +48,10 @@ type MapType struct {
-	Key       interface{} // usually a string
-	ValueType Type
+	Key          interface{} // usually a string
+	ValueType    Type
+	DefaultValue interface{}
```
`ValueType` has access to its source which contains the original node (and therefore the default value)...

Was that in the mix when we considered introducing `DefaultValue`?

joaopapereira replied:
Yes indeed `ValueType` is of the type `Type` (lol) but we would have to cast it to the correct type in line 133 of this file in order to retrieve the value from this new type.
To simplify that logic, we decided to add another field that contains the raw value we want to attach

joaopapereira replied:
If you know of a better way let me know and I am happy to update it 😃

pivotaljohn commented on go.sum:
```diff
@@ -119,6 +119,7 @@ github.com/spf13/pflag v1.0.3/go.mod h1:DYY7MBk1bdzusC3SYhjObp+wFpr4gzcvqqNjLnIn
 github.com/spf13/pflag v1.0.5 h1:iy+VFUOCP1a+8yFto/drg2CJ5u0yRoB7fZw3DKv/JXA=
 github.com/spf13/pflag v1.0.5/go.mod h1:McXfInJRrz4CZXVZOBLb0bTZqETkiAhM9Iw0y3An2Bg=
 github.com/spf13/viper v1.4.0/go.mod h1:PTJ7Z/lr49W6bUbkmS1V3by4uWynFiR9p7+dSq/yZzE=
+github.com/stretchr/objx v0.1.1 h1:2vfRuCMp5sSVIDSqO8oNnWJq7mPa6KVP3iPIwFBuy8A=
```
I'm still learning about Go modules... what does it mean for the version number not to change but the checksum does not?

joaopapereira replied:
I added a package to do a check, and that updated the `go.sum`. I removed this line and everything is working now

pivotaljohn commented on pkg/yamlmeta/type.go:
```diff
@@ -103,18 +105,48 @@ func (t *MapType) AssignTypeTo(typeable Typeable) (chk TypeCheck) {
+		}
+		childCheck := item.AssignTypeTo(val)
+		chk.Violations = append(chk.Violations, childCheck.Violations...)
+		mapNode.AddValue(val) // TODO: How might we surface this error?
```
Open `TODO:`.

I think this is an unexpected situation. Are there cases where applying the default value _could_ result in a **type-check** violation (as opposed to a validation violation)?

Given that considering, what should we do, here?

pivotaljohn reviewed (changes requested):
Looks really good, @joaopapereira.

Some nip and tuck still suggestions, below.

cppforlife commented on pkg/yamlmeta/type.go:
```diff
@@ -103,18 +105,48 @@ func (t *MapType) AssignTypeTo(typeable Typeable) (chk TypeCheck) {
+		}
+		childCheck := item.AssignTypeTo(val)
+		chk.Violations = append(chk.Violations, childCheck.Violations...)
+		mapNode.AddValue(val) // TODO: How might we surface this error?
```
if we are not propagating error up then we should panic. definitely do not want to swallow.

joaopapereira replied:
To get an error we would need to pass a value to the method `AddValue` that is not of the type `*MapItem`. This cannot happen because we create the variable `val` 7 lines above this invocation.
Nevertheless, we can always add here a panic, not really a fan of panic in mid execution 😢  that was why I added a TODO so we can discuss the best option

cppforlife replied:
> Nevertheless, we can always add here a panic, not really a fan of panic in mid execution

`panic("Unreachable")` is a very common pattern to assert on internal constraints.

joaopapereira replied:
done

joaopapereira committed:
Add support for default values fill in

When a value is missing in the data file but present in the schema, ytt
will use this value as the default and it will populate the map
structure with it.

joaopapereira committed:
Changes from code review

joaopapereira committed:
Add unreachable panic for unreachable error

joaopapereira committed:
Remove go.sum change

cppforlife commented on pkg/yamlmeta/type.go:
```diff
@@ -103,18 +105,52 @@ func (t *MapType) AssignTypeTo(typeable Typeable) (chk TypeCheck) {
+		chk.Violations = append(chk.Violations, childCheck.Violations...)
+		err := mapNode.AddValue(val)
+		if err != nil {
+			panic("Unreachable")
```
i would go with something like this actually `fmt.Sprintf("Internal inconsistency: adding map item: %s", err)`

joaopapereira replied:
fixed

cppforlife commented on pkg/cmd/template/schema_test.go:
```diff
@@ -361,3 +429,27 @@ rendered: true`
 	}
 }
+
+func diffText(left, right string) string {
```
seems like this got copied from elsewhere. lets pull that out into difflib.

joaopapereira replied:
extracted and updated the library

joaopapereira committed:
Correct the panic message

joaopapereira committed:
Update difflib to use PPDiff function

Refactor some tests to use PPDiff
Create helper method to abstract the call to PPDiff

pivotaljohn commented on pkg/yamlmeta/type.go:
```diff
@@ -137,7 +137,7 @@ func (t *MapType) applySchemaDefaults(foundKeys []interface{}, chk TypeCheck, ma
 		err := mapNode.AddValue(val)
 		if err != nil {
-			panic("Unreachable")
+			panic(fmt.Sprintf("Internal inconsistency: adding map item: %s", err))
```
Improved! 👍🏼 

In general, when I'm faced with reporting particularly rare/unexpected situations, I try to report as much context as I can.  So, I scan around the function that is going to emit this panic and look for values that are essential to the failed operation and make sure they are either included in the Error message, or report it myself.  For example, is `val` in the message of `err`? And what's the state of `mapNode`?

I think this is okay, for now, but would be worth the little extra effort to make sure those surprising moments come more wrapped in a bow.
